### PR TITLE
Add support for partial Zarr download and upload

### PR DESCRIPTION
**Pull request by yarikoptic**
## Summary

Design and implementation for partial zarr download and upload support, addressing #1462, #1474, and related archive issues.

- Closes #1596 as AFAIK supersedes it

The PR covers five areas:

1. **`--zarr TYPE:PATTERN` filtering** for `dandi download` — glob, path, and regex filters for selecting entries within zarr assets, with a `metadata` alias for common zarr metadata files
2. **URL parsing with zarr boundary detection** — `AssetZarrEntryURL` to handle URLs like `dandi://dandi/000108/.../file.ome.zarr/0/0/0`
3. **`--zarr-mode {full, patch}`** for `dandi upload` — patch mode uploads changed files without deleting remote files absent locally
4. **Checksums and manifests** — documents that per-directory checksums are computed hierarchically by the `zarr_checksum` library but are NOT persisted (only the root digest is stored in the DB); legacy `.checksum` files exist on S3 at `zarr-checksums/` for ~72% of older zarrs but are orphaned since Dec 2022
5. **`dandi ls` for zarr contents** — listing files within a zarr asset when URL points into a zarr

### Key findings from investigation

- The `zarr_checksum` algorithm IS hierarchical (Merkle tree, bottom-up via `ZarrChecksumTree`)
- The archive's `ingest_zarr_archive` task computes checksums **entirely in memory** and stores only the root digest
- Per-directory `.checksum` files on S3 (`zarr-checksums/` prefix) were written by `ZarrChecksumFileUpdater`, removed in dandi-archive PRs #1390/#1395 (Dec 2022). Legacy files remain for older zarrs but no API exposes them
- Subtree checksum verification is not possible today without recomputation from file ETags

### New files
- `dandi/zarr_filter.py` — `ZarrFilter` dataclass with glob/path/regex matching, `parse_zarr_filter()`, `make_zarr_entry_filter()`, `ZARR_FILTER_ALIASES` (includes `metadata` alias)
- `dandi/tests/test_zarr_filter.py` — 52 unit tests covering all filter types, parsing, aliases, edge cases, invalid regex validation

### Modified files
- `dandi/dandiarchive.py` — `split_zarr_location()`, `AssetZarrEntryURL` class, updated `parse_dandi_url()` to detect zarr boundaries
- `dandi/download.py` — `zarr_filters` parameter on `download()`, filter threading through `Downloader` and `_download_zarr()`, skip deletion/checksum when filter active
- `dandi/files/zarr.py` — `zarr_mode: Literal["full", "patch"]` on `iter_upload()`, patch mode skips remote file deletion and client-side checksum verification
- `dandi/upload.py` — `zarr_mode` parameter on `upload()`, conditional pass-through for `ZarrAsset`
- `dandi/cli/cmd_download.py` — `--zarr` click option (multiple, OR logic)
- `dandi/cli/cmd_upload.py` — `--zarr-mode` click option
- `dandi/cli/cmd_ls.py` — list zarr entries when URL is `AssetZarrEntryURL`
- `dandi/cli/tests/test_download.py` — updated mock expectations for new `zarr_filters` parameter

### Integration tests
- `dandi/tests/test_download.py` — 6 tests (glob filter, metadata alias, no-delete, path filter, nonexistent filter, sync conflict)
- `dandi/tests/test_upload.py` — 3 tests (patch no-delete, full delete, patch updates)
- `dandi/tests/test_dandiarchive.py` — 3 URL parsing cases + 8 `split_zarr_location` cases

## Review checklist

Please review the design at [`doc/design/partial-zarr.md`](doc/design/partial-zarr.md) and comment on:

- [x] `--zarr TYPE:PATTERN` syntax — is the filter approach right? Are glob/path/regex the right types?
- [x] `metadata` alias expansion — does `glob:**/.z*` + `glob:**/zarr.json` + `glob:**/.zmetadata` cover all cases?
- [x] `--zarr-mode patch` semantics — is "upload without deleting" the right default for patch? ~~Should subtree cleanup happen?~~
- [x] URL parsing — is `AssetZarrEntryURL` with zarr boundary detection the right approach?
- [x] Checksum strategy — relying on per-file ETags for partial ops, deferring subtree checksums to future manifests
- [x] Open questions in the doc (OR composition chosen, `--sync` conflict raises error, server-side glob deferred)
- [ ] Should legacy `zarr-checksums/` files on S3 be cleaned up as part of this or separately?

## TODO

- [x] Implement `dandi/zarr_filter.py` — filter parsing and matching
- [x] Implement `AssetZarrEntryURL` and `split_zarr_location()` in `dandi/dandiarchive.py`
- [x] Add `--zarr` option to `dandi download` CLI
- [x] Modify `_download_zarr()` for partial download support
- [x] Add `--zarr-mode` option to `dandi upload` CLI
- [x] Implement patch mode in `iter_upload()` (`dandi/files/zarr.py`)
- [x] Thread `zarr_mode` through `dandi/upload.py`
- [x] `dandi ls` zarr contents support
- [x] Unit tests for `zarr_filter.py` (52 tests)
- [x] Unit tests for URL parsing and `split_zarr_location` (11 test cases)
- [x] Integration tests for download filtering (6 tests)
- [x] Integration tests for upload patch mode (3 tests)
- [x] Fix existing CLI mock test expectations
- [x] Code review fixes (regex pre-compilation, `Literal` typing, `**` collapse, redundant override removal)
- [ ] Improve reporting difference per @kabilar [comment below](https://github.com/dandi/dandi-cli/pull/1816#issuecomment-4919027625) for skipped-vs-updated zarr
- [ ] Run integration tests against docker-compose archive instance
- [ ] Coordinate with dandi-archive on manifest design (#2702) for future subtree checksum support

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: doc/design/partial-zarr.md [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+# 1. Download just the metadata files from a zarr
+dandi download --zarr glob:'**/.z*' --zarr glob:'**/zarr.json' \
+    dandi://dandi/001289/rawdata/.../PC.ome.zarr
+
+# 2. Edit .zattrs locally
+vim PC.ome.zarr/.zattrs
+
+# 3. Upload changes without deleting remote data chunks
+dandi upload --zarr-mode patch
+```
```
Comment (kabilar) [inline doc/design/partial-zarr.md]: Thanks @yarikoptic.  An additional use case would be adding shards to an existing Zarr.  For example, we have 355+ TB from a subject that were are stitching together into a single Zarr.  We don't have enough space locally so will need to upload a shard, delete the local copy, and repeat the process.

cc @satra @ayendiki

File: doc/design/partial-zarr.md [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+| Full upload | Yes (via S3 Content-MD5) | N/A | Yes (client vs server) |
+| Patch upload | Yes (via S3 Content-MD5) | No (future: via manifest) | No (skipped) |
+
+---
+
+## Part 5: `dandi ls` for Zarr Contents
+
+See separate implementation task.  Summary: allow `dandi ls` to list files
+within a Zarr asset when given a Zarr URL, using `asset.iterfiles(prefix=...)`.
+Reuses `AssetZarrEntryURL` from Part 2.
```
Comment (kabilar) [inline doc/design/partial-zarr.md]:
```suggestion
This will be implemented as a separate pull request.  Summary: allow `dandi ls` to list files
within a Zarr asset when given a Zarr URL, using `asset.iterfiles(prefix=...)`.
Reuses `AssetZarrEntryURL` from Part 2.
```

File: doc/design/partial-zarr.md [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+| `dandi/download.py` | `_download_zarr()` filtering, `Downloader` plumbing |
+| `dandi/upload.py` | Pass `zarr_mode` through |
+| `dandi/files/zarr.py` | `iter_upload()` patch mode support |
+| `dandi/cli/cmd_ls.py` | Zarr contents listing (separate PR) |
+
+## Open Questions
+
+1. **`--zarr` AND vs OR**: Multiple `--zarr` options use OR.  Should we support
+   AND composition (e.g., `--zarr path:0/1 --zarr-and glob:'**/.zattrs'`)?
+   For now, OR suffices; AND can be added later.
```
Comment (kabilar) [inline doc/design/partial-zarr.md:495]: Just having the OR operator is fine with me.

File: doc/design/partial-zarr.md [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+   happens client-side after fetching entries.  For large zarrs with millions of
+   entries, this could be slow.  A future server API enhancement for
+   entry-level glob would help.  For now, `path:` filters should be preferred
+   for large zarrs to minimize data transfer.
+
+3. **Interaction with `--sync`**: When `dandi download --zarr ... --sync` is
+   used, should `--sync` only delete local zarr entries that match the filter
+   but aren't remote?  Or should `--sync` be disallowed with `--zarr`?
+   Recommendation: disallow `--sync` with `--zarr` initially, as the semantics
+   are ambiguous.
```
Comment (kabilar) [inline doc/design/partial-zarr.md:508]: As mentioned, since the semantics could be confusing, I would suggest disallowing `--sync` with `--zarr` for `dandi download`.

File: doc/design/partial-zarr.md [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+1. **`--zarr` AND vs OR**: Multiple `--zarr` options use OR.  Should we support
+   AND composition (e.g., `--zarr path:0/1 --zarr-and glob:'**/.zattrs'`)?
+   For now, OR suffices; AND can be added later.
+
+2. **Server-side glob for zarr entries**: Currently the server only supports
+   `prefix` filtering on zarr entries (not glob).  Glob filtering therefore
+   happens client-side after fetching entries.  For large zarrs with millions of
+   entries, this could be slow.  A future server API enhancement for
+   entry-level glob would help.  For now, `path:` filters should be preferred
+   for large zarrs to minimize data transfer.
```
Comment (kabilar) [inline doc/design/partial-zarr.md:502]: This works.  We are using Zarr v3 so glob filtering on the client-side shouldn't be an issue.

File: doc/design/partial-zarr.md [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+| `dandi/consts.py` | `ZARR_METADATA_FILENAMES` constants (used by `metadata` alias) |
+| `dandi/dandiarchive.py` | `AssetZarrEntryURL`, `split_zarr_location()`, modify `parse_dandi_url()` |
+| `dandi/cli/cmd_download.py` | `--zarr` option |
+| `dandi/cli/cmd_upload.py` | `--zarr-mode` option |
+| `dandi/download.py` | `_download_zarr()` filtering, `Downloader` plumbing |
+| `dandi/upload.py` | Pass `zarr_mode` through |
+| `dandi/files/zarr.py` | `iter_upload()` patch mode support |
+| `dandi/cli/cmd_ls.py` | Zarr contents listing (separate PR) |
+
+## Open Questions
```
Comment (kabilar) [inline doc/design/partial-zarr.md:491]: I agree with all of the proposed solutions for the open questions.

File: doc/design/partial-zarr.md
Comment (kabilar) [inline doc/design/partial-zarr.md:1]: Thanks @yarikoptic.  This design looks good and works for our two use cases: 1. updating Zarr metadata, and 2. adding chunks/shards to an existing Zarr.

Comment (kabilar) [reply doc/design/partial-zarr.md:1]: Hi @yarikoptic, @ayendiki and I were discussing this feature today.  What would be the next steps to push this forward?  Thanks.

File: doc/design/partial-zarr.md (hunk 4 above)
Comment (yarikoptic) [reply doc/design/partial-zarr.md:508]: coolio, that is already implemented above https://github.com/dandi/dandi-cli/pull/1816/changes#diff-a0dcf236f5c18d66953b7985098c75c21bbc9dd6b7e6e7063d5d38c2044f6697R143

File: doc/design/partial-zarr.md (hunk 1 above)
Comment (yarikoptic) [reply doc/design/partial-zarr.md]: do you see currently proposed/implemented interface be sufficient for this use case?

File: doc/design/partial-zarr.md
Comment (yarikoptic) [reply doc/design/partial-zarr.md:1]: have you tried it on your data/use cases ?

if so, and it would be sufficient and convenient enough in terms of API/CLI -- we can refresh branch/address conflicts and seek more reviews and merge.  If not (something else should be implemented) -- state the needs

Comment (kabilar) [reply doc/design/partial-zarr.md:1]: I didn't realize this was ready for testing since it is in `Draft` mode.  Will test and let you know, thanks.

**Comment by kabilar** [comment]
Hi @yarikoptic, thanks for this pull request.  I was able to add objects to an existing zarr, delete objects, and modify metadata (i.e. `zarr.json`).

I just have some minor feedback on the terminal output.

When I run `dandi upload` for a dandiset with a zarr that is the same between the local storage and dandi, the `STATUS` is `done` instead of `skipped` and the `MESSAGE` is `exists - reuploading` instead of `file exists`: 

```
PATH                                              SIZE      ERRORS PROGRESS STATUS                         MESSAGE                  
dandiset.yaml                                     1.8 kB                    skipped                        should be edited online  
.bidsignore                                       13 Bytes    0             skipped                        file exists              
dataset_description.json                          404 Bytes   0             skipped                        file exists              
...atives/dandi-cli-1816-test/HG9_Z1_Y49.nii.zarr 5.9 GB      0             done                           exists - reuploading     
Summary:                                          5.9 GB                    3 skipped                      1 should be edited online
                                                                            1 done                         2 file exists            
                                                                                                           1 exists - reuploading 
```

And then when I modify the zarr locally and run `dandi upload`, the `STATUS` and `MESSAGE` are the same:

```
PATH                                             SIZE      ERRORS   PROGRESS STATUS                        MESSAGE                  
dandiset.yaml                                    1.8 kB                      skipped                       should be edited online  
.bidsignore                                      13 Bytes    0               skipped                       file exists              
dataset_description.json                         404 Bytes   0               skipped                       file exists              
...tives/dandi-cli-1816-test/HG9_Z1_Y49.nii.zarr 6.6 GB      0          100% done                          exists - reuploading     
Summary:                                         6.6 GB             3.3 MB/s 3 skipped                     1 should be edited online
                                                                             1 done                        2 file exists            
                                                                                                           1 exists - reuploading 
```

So I would propose the following changes to the terminal output:
1. If the zarr exists on DANDI and is the same as the local version, the `STATUS` should be `skipped` and `MESSAGE` should be `file exists`.
2. If the zarr exists on DANDI and is different than the local version, the `MESSAGE` should be more specific to what is happening to the zarr.  For example:
    - `file exists - uploading additional objects (2 GB)`
    - `file exists - deleting objects (2 GB)`
    - `file exists - modifying objects (2 GB)`
    - Note that I have also added `(2 GB)` in the example to denote how many of the objects in the zarr are being uploaded/deleted/modified, since the `SIZE` is that of the entire Zarr tree.

**Comment by kabilar** [comment]
Just explicitly noting that `dandi download --zarr metadata <path>` and `dandi upload --zarr-mode patch` both work as intended.  

For `dandi download --zarr metadata dandi://dandi/001769@draft/derivatives/dandi-cli-1816-test/HG9_Z1_Y49.nii.zarr`, this downloaded all the `zarr.json` files in the Zarr tree.

For `dandi upload --zarr-mode patch`, only modified local files were uploaded and remote files (that are absent locally) were not deleted.

File: doc/design/partial-zarr.md (hunk 1 above)
Comment (kabilar) [reply doc/design/partial-zarr.md]: Yes, thanks.  Should we add this as a use case in the design doc?

Comment (yarikoptic) [reply doc/design/partial-zarr.md]: we could -- just push it to this branch or send a PR against it if would like a review

Comment (kabilar) [reply doc/design/partial-zarr.md]:
```suggestion
## Use Cases

### Use Case 1 - Update metadata

1. Download just the metadata files from a Zarr.
  `dandi download --zarr glob:'**/.z*' --zarr glob:'**/zarr.json' \
      dandi://dandi/001289/rawdata/.../PC.ome.zarr`
2. Edit metadata locally.
  `vim PC.ome.zarr/zarr.json`
3. Upload changes without deleting remote data chunks.
  `dandi upload --zarr-mode patch`

### Use Case 2 - Incrementally add shards to an existing Zarr

We are generating a very large dataset for which there is not enough space to house the entire dataset locally.

1. Add shard(s) locally to a Zarr that exists on DANDI.
2. Upload shard(s) without deleting remote Zarr objects (that don't exist locally).
3. Remove local shard(s) and repeat the process.
```

**Comment by kabilar** [comment]
I am re-running the failed CI workflows, because I am not sure why they are only failing for ubuntu python>3.10.

File: dandi/upload.py [hunk 7]
```diff
@@ -97,6 +97,14 @@ class UploadValidation(StrEnum):
     IGNORE = "ignore"
 
 
+class ZarrMode(str, Enum):
+    FULL = "full"
+    PATCH = "patch"
+
+    def __str__(self) -> str:
+        return self.value
+
```
Comment (kabilar) [inline dandi/upload.py]: `claude-haiku-4.5` says that the following should fix the failing tests.

```suggestion
class ZarrMode(StrEnum):
    FULL = "full"
    PATCH = "patch"


```

**Comment by CodyCBakerPhD** [comment]
> I am not sure why they are only failing for ubuntu python>3.10.

Fixed it for you - simple discrepancy for v2/v3 metadata structure in the new glob test

**Comment by kabilar** [comment]
> > I am not sure why they are only failing for ubuntu python>3.10.
> 
> Fixed it for you - simple discrepancy for v2/v3 metadata structure in the new glob test

Awesome, thank you

File: doc/design/partial-zarr.md (hunk 1 above)
Comment (yarikoptic-gitmate) [reply doc/design/partial-zarr.md]: Added as Use Case 2 in 1a522912 ("doc+test: partial-zarr Use Case 2 (incremental shard upload)"). Thanks for the concrete motivation from your 355+ TB dataset -- I also added a regression test (`test_upload_zarr_patch_mode_incremental_shards`) that prunes a Zarr down to metadata files, drops in a synthetic shard, and verifies that patch-mode upload adds the new shard while leaving remote-only entries intact.

File: doc/design/partial-zarr.md (hunk 2 above)
Comment (yarikoptic-gitmate) [reply doc/design/partial-zarr.md]: Applied in d43f3639 ("DOC: Minor tune up in phrasing of the design doc").

File: dandi/upload.py (hunk 7 above)
Comment (yarikoptic-gitmate) [reply dandi/upload.py]: Applied in 815fa198 (@CodyCBakerPhD, co-authored with @kabilar) -- switched `ZarrMode(str, Enum)` to `ZarrMode(StrEnum)` and dropped the redundant `__str__`.

**Comment by yarikoptic-gitmate** [comment]
Thanks @kabilar for the detailed report and the concrete before/after mockups -- both are real bugs in what `dandi upload` prints for Zarr assets:

1. **Unchanged Zarr misreported as \`done / exists - reuploading\`** -- should be \`skipped / file exists\`, matching the non-Zarr case. This is a longstanding cosmetic issue in \`dandi/files/zarr.py::iter_upload\` (the same-hash short-circuit yields a \`done\` status token instead of \`skipped\`).
2. **Modified Zarr message is not descriptive** -- your proposed vocabulary (\`file exists - uploading additional objects (N GB)\`, \`... deleting objects (N GB)\`, \`... modifying objects (N GB)\`) is a nice improvement over \`exists - reuploading\`; adding the affected-bytes suffix is genuinely useful given \`SIZE\` reflects the whole Zarr tree.

Both are orthogonal to the partial-zarr download/upload plumbing this PR introduces (they would apply just as much on \`master\` today), so I'd like to keep the scope of #1816 focused on landing partial-zarr and track your terminal-output feedback in a follow-up issue. Would that work for you? I can open the follow-up issue and reference this comment.

**Comment by yarikoptic** [comment]
as agreed at weekly meeting, here is a dedicated issue
- https://github.com/dandi/dandi-cli/issues/1893
to follow up with

**Comment by yarikoptic** [comment]
@kabilar if overall satisfied etc, approve to be merged ;-)

File: dandi/cli/cmd_download.py [hunk 8]
```diff
@@ -132,6 +132,17 @@
     "With 'ask' (the default when --sync is passed without a value), prompt before "
     "deleting. With 'do', delete without prompting.",
 )
+@click.option(
+    "--zarr",
```
Comment (satra) [inline dandi/cli/cmd_download.py:136]: should this be an explicit option or inferred from metadata like neuroglancer does? or a hint when the user/agent knows specifically to do so.

File: dandi/cli/cmd_upload.py [hunk 9]
```diff
@@ -57,6 +57,16 @@
     default="require",
     show_default=True,
 )
+@click.option(
+    "--zarr-mode",
```
Comment (satra) [inline dandi/cli/cmd_upload.py:62]: same comment here. we are increasing the CLI surface and should ask if these should be there or inferred.

File: dandi/dandiarchive.py [hunk 10]
```diff
@@ -446,6 +447,72 @@ def get_assets(
...
+    """Split a location into ``(asset_path, zarr_subpath)`` if it crosses a zarr boundary.
+
+    Scans path components for segments ending with a Zarr extension
+    (``.zarr``, ``.ngff``).  If found **and** there are remaining components
+    after the boundary, returns the split; otherwise returns ``None``.
+
+    Parameters
+    ----------
+    location : str
+        A POSIX-style path, e.g. ``"sub-1/file.ome.zarr/0/0/0"``.
```
Comment (satra) [inline dandi/dandiarchive.py:460]: zarr doesn't have to have a `/` separator. a local zarr could have been generated with a different separator (e.g. `zarr/0.0.0`) or sharded.

File: dandi/zarr_filter.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+Provides filter parsing, matching, and predefined aliases for selecting
+subsets of entries within Zarr assets (e.g., only metadata files).
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from fnmatch import fnmatchcase
+import re
+from typing import Callable, Literal
```
Comment (satra) [inline dandi/zarr_filter.py:13]: this file seems to do a lot of things without using a zarr or ome-zarr library, which could help deal with versioning, chunking, and sharding, as appropriate. and for ome-zarr elements related to metadata changes.

**Comment by kabilar** [comment]
Hi @yarikoptic, just following up from our discussion last week.  Is this ready again for testing?  Thanks.

File: dandi/cli/cmd_download.py (hunk 8 above)
Comment (yarikoptic) [reply dandi/cli/cmd_download.py:136]: we infer already based on extensions IIRC; this is just an a new option for filter... I guess we might better call it that `--zarr-filter` (with `--z-f` shortcut)

File: dandi/cli/cmd_upload.py (hunk 9 above)
Comment (yarikoptic) [reply dandi/cli/cmd_upload.py:62]: and the same reply here: it is not about the fact that we are dealing with zarr, but rather an option specific if working with zarr(s).  Let's also add shortcut `--z-m`

File: dandi/dandiarchive.py (hunk 10 above)
Comment (yarikoptic) [reply dandi/dandiarchive.py:460]: if sharded -- don't we upload it "as is" with all its shardiness?

it is a new info for me and I guess refers to `dimension_separator` in v2 and now `chunk_key_encoding` in v3? (docs: https://zarr.readthedocs.io/en/v3.1.0/api/zarr/)

good to know but I wonder if that is an issue bigger than this "docstring fix" e.g.

- are we here to map here from `.` to `/` and it is client's job
- is our dandi-archive code capable of handling those correctly?
- does https://github.com/dandi/zarr_checksum ?

likely worth a concerted effort to support such ones (unless I am wrong and we already do). Pragmatically, does any of the actively working with DANDI groups use alternative separators @satra @kabilar ?  I think worth a separate issue, and if you see what to change here specifically -- please leave a suggestion @satra

File: dandi/zarr_filter.py (hunk 11 above)
Comment (yarikoptic) [reply dandi/zarr_filter.py:13]: we could potentially later add some more "semantic driven" filtering on what to download or upload, but here it is not the goal, we just implement based on basic `glob` and metadata matching (agnostic whether it is ome or not) and we do not track any heavy dependencies in.   Since we have those matching namespaces (e.g. `glob:`) we will be able to extend later with some `ome-zarr:` if needed .  But I would not blow it up even more for this one

**Review by yarikoptic — COMMENTED** [review]
Replies to @satra's replies. 

Also remember "perfect is an enemy of the good"!

**Comment by yarikoptic** [comment]
ok, if no objections voiced, I will merge/release it tomorrow so it could be attempted to be used "in production", and contributions will be welcome to enhance with additional ways.

**Review by yarikoptic-gitmate — APPROVED** [review]
Since @kabilar confirmed this PR helping out in his use-cases, and no further objections were voiced, let's proceed!
